_posts/2021-07-05-newsletter.md

```diff
@@ -0,0 +1,199 @@
+
+- completion,
+- hover/tooltips,
+- go to definition,
+- show/go to references,
+- show method signatures,
+- rename,
+- code actions (automatic formatting, organize imports, ...),
+
+and more.
+
+The idea is to separate these features into an editor-independent but language-specific server and a language-independent but editor-specific client, which communicate via the Language Server Protocol over RPC. (It should be pointed out that not every server implements every feature, and the quality of the responses can vary wildly. The "reference implementation" in VS Code also often adds non-standard features that are not covered by the LSP itself).
+
+Neovim 0.5 provides an LSP client written (mostly) in Lua that provides a highly configurable and extensible way of accessing these features. It does not aim at competing with more feature-rich and "out-of-the-box" plugins like [CoC.nvim](https://github.com/neoclide/coc.nvim) but is meant to be tailored to your preferences (while still being useable with reasonable defaults). For an overview, see TJ DeVries' [Vimconf.live presentation](https://www.youtube.com/watch?v=C9X5VF9ASac) and his [shorter video](https://www.youtube.com/watch?v=ArwDgvYEZYk).
+
+For many language servers, [Nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) already provides the necessary configuration to set everything up easily. Alternatively, some languages also have specific LSP plugins that provide a more integrated setup, e.g., for [Java](https://github.com/mfussenegger/nvim-jdtls) and [Scala](https://github.com/ckipp01/nvim-metals).
+
+To learn more about LSP and how to use it in Neovim, visit [Nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) and read [`:h lsp`](https://neovim.io/doc/user/lsp.html).
+
+Expect more work on LSP to provide more configuration options and full coverage of LSP version 3.16 (including semantic highlighting).
+
+## Tree-sitter
+
+Neovim 0.5 adds **experimental** support for tree-sitter, a library that parses a piece of code into a syntax tree in an incremental and error-resilient way; this means that reparsing that code after an edit is very fast, and parsing error due to, e.g., typos remain localized and do not break parsing further down. This tree can then be efficiently queried to obtain syntax information about the code. This allows for improved and/or faster
+
+- syntax highlighting,
+- code navigation,
+- refactoring,
+- text objects and motions,
+- searching and replacing,
+
+and more.
+
+In addition, tree-sitter allows "injecting" trees for other languages, which makes it easy to support, e.g., highlighting code blocks in Markdown documents. To learn more about tree-sitter, watch [Tree-sitter - A new parsing system for programming tools - Max Brunsfield](https://www.youtube.com/watch?v=Jes3bD6P0To&feature=youtu.be&t=232).
```

Comment:
`which makes it easy to support files that contain code in several languages`

Reply:
Yes, that's better (since specifically Markdown has issues :P)